satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,323 @@
+	defer mon.Task()(&ctx)(&err)
+
+	newNodeSelection := `SELECT last_net, id, address, last_ip_port, true FROM nodes`
+	reputableNodeSelection := `SELECT last_net, id, address, last_ip_port, false FROM nodes`
+
+	if criteria.DistinctIP {
+		newNodeSelection = `SELECT DISTINCT ON (last_net) last_net, id, address, last_ip_port, true FROM nodes`
+		reputableNodeSelection = `SELECT DISTINCT ON (last_net) last_net, id, address, last_ip_port, false FROM nodes`
+	}
+
+	newNodesCondition, err := buildConditions(ctx, criteria, excludedIDs, excludedNetworks, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	reputableNodesCondition, err := buildConditions(ctx, criteria, excludedIDs, excludedNetworks, false)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var newNodeQuery, reputableNodeQuery query
+	newNodeQuery = assemble(partialQuery{selection: newNodeSelection, condition: newNodesCondition, limit: newNodeCount})
+	if criteria.DistinctIP {
+		newNodeQuery = assemble(partialQuery{selection: newNodeSelection, condition: newNodesCondition, orderBy: "last_net", distinct: true, limit: newNodeCount})
+		newNodeQuery = selectRandomFrom(newNodeQuery, "filteredcandidates", newNodeCount)
+	}
+
+	reputableNodeQuery = assemble(partialQuery{selection: reputableNodeSelection, condition: reputableNodesCondition, limit: reputableNodeCount})
+	if criteria.DistinctIP {
+		reputableNodeQuery = assemble(partialQuery{selection: reputableNodeSelection, condition: reputableNodesCondition, orderBy: "last_net", distinct: true, limit: reputableNodeCount})
+		reputableNodeQuery = selectRandomFrom(reputableNodeQuery, "filteredcandidates", reputableNodeCount)
+	}
+
+	query := union(newNodeQuery, reputableNodeQuery)
+
+	rows, err := cache.db.Query(ctx, cache.db.Rebind(query.query), query.args...)
+	if err != nil {
+		return nil, nil, Error.Wrap(err)
+	}
+	defer func() { err = errs.Combine(err, rows.Close()) }()
+
+	for rows.Next() {
+		var node overlay.SelectedNode
+		node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+		var lastIPPort sql.NullString
+		var isNew bool
+		err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort, &isNew)
+		if err != nil {
+			return nil, nil, err
+		}
+		if lastIPPort.Valid {
+			node.LastIPPort = lastIPPort.String
+		}
+		if isNew {
+			newNodes = append(newNodes, &node)
+		} else {
+			reputableNodes = append(reputableNodes, &node)
+		}
+		if len(newNodes)+len(reputableNodes) == reputableNodeCount+newNodeCount {
+			break
+		}
+	}
+	return reputableNodes, newNodes, nil
+}
+
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string, isNewNodeQuery bool) (condition, error) {
+	var conds conditions
+	if isNewNodeQuery {
+		conds.add(
+			"total_audit_count < ?",
```

Comment:
we should probably add `OR total_uptime_count < ?` so that it is consistent with the reputation criteria for vetted nodes
(@jenlij )

Reply:
so the reason I left this out is because in the original code at the satellite/overlay/service.go level, total uptime count isn't even passed in as a selection criteria: https://github.com/storj/storj/blob/master/satellite/overlay/service.go#L315

so right now, the `total_uptime_count` will always be zero, which means that only the audit count check will matter at this filter: https://github.com/storj/storj/blob/master/satellite/satellitedb/overlaycache.go#L102

Reply:
Yeah, I am aware of the inconsistency in the original code - I was discussing it with @jenlij yesterday. But I think it would be better to include total_uptime_count, even if it is zero just so we have consistency.

Either way, Jen is working on a change to have a `vetted` flag, which will be used instead of comparing against total_audit_count and total_uptime_count anyway, so it's not a big deal either way. I don't consider this to be an approval-blocker.